scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/edf/EDFPolicyExtended.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * ProActive Parallel Suite(TM):
+ * The Open Source library for parallel and distributed
+ * Workflows & Scheduling, Orchestration, Cloud Automation
+ * and Big Data Analysis on Enterprise Grids & Clouds.
+ *
+ * Copyright (c) 2007 - 2017 ActiveEon
+ * Contact: contact@activeeon.com
+ *
+ * This library is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Affero General Public License
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.scheduler.policy.edf;
+
+import java.time.Duration;
+import java.util.Calendar;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.ow2.proactive.scheduler.common.JobDescriptor;
+import org.ow2.proactive.scheduler.common.job.JobPriority;
+import org.ow2.proactive.scheduler.descriptor.EligibleTaskDescriptor;
+import org.ow2.proactive.scheduler.descriptor.JobDescriptorImpl;
+import org.ow2.proactive.scheduler.job.InternalJob;
+import org.ow2.proactive.scheduler.policy.ExtendedSchedulerPolicy;
+
+
+/**
+ * Early Deadline First Policy sorts jobs based on:
+ * - job priorities (from highest to lowest)
+ * - job deadline (those that have deadline overtake those without deadline)
+ *
+ * Among the job with the dealine (and having same priority) we distinguish
+ * them by the fact if some task of the job is already scheduled (i.g. job has
+ * a startingTime) or not. Those that started has a priority to the not yet started jobs.
+ * Among the jobs that are already started we sort them by startTime.
+ * If job is not started, we sort them by `effectiveDeadline - (now() + expectedTime`.
+ * If deadline is absolute then effectiveDeadline equals to deadline.
+ * If deadline is relative then effectiveDeadline is calculated as (now() + deadline)
+ *
+ * Among jobs without deadline (and the same priority), we sort them by submission date.
+ *
+ * @author The ProActive Team
+ * @since ProActive Scheduling 8.3
+ */
+public class EDFPolicyExtended extends ExtendedSchedulerPolicy {
+
+    private static final Date MAXIMUM_DATE = new Date(Long.MAX_VALUE);
+
+    @Override
+    public LinkedList<EligibleTaskDescriptor> getOrderedTasks(List<JobDescriptor> jobs) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') The return type of this method should be an interface such as "List" rather than the implementation "LinkedList". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1319)